import time
from fastapi import FastAPI, Request, Response
import subprocess
import os

app = FastAPI()

@app.get("/")
async def root():
    return {"message": "Server is Ok funciona PORRA!"}

@app.post("/webhook")
async def webhook(request: Request):
    print("‚úÖ Webhook recebido. Executando git pull...")

    try:
        # Executando o comando git pull
        output = subprocess.check_output(['git', '-C', os.getcwd(), 'pull'], stderr=subprocess.STDOUT)
        print("üì¶ Git pull output:\n", output.decode())

        # Espera de 3 segundos para garantir que as altera√ß√µes foram aplicadas
        print("‚è≥ Aguardando para garantir que as altera√ß√µes sejam aplicadas...")
        time.sleep(3)  # Pode ajustar o tempo conforme necess√°rio

        print("‚ôªÔ∏è Reiniciando o servidor FastAPI com o script de rein√≠cio...")

        # Chama o script restart.sh para reiniciar o servidor
        result = subprocess.run(['bash', 'restart.sh'], cwd=os.getcwd(), text=True, capture_output=True)

        if result.returncode != 0:
            print(f"‚ùå Erro ao reiniciar servidor. C√≥digo de erro: {result.returncode}")
            print(f"Detalhes do erro: {result.stderr}")
            return Response(content="Falha ao reiniciar o servidor", status_code=500)

        print("‚úÖ Servidor reiniciado com sucesso.")

        return Response(content="Success and restarted!", status_code=200)

    except subprocess.CalledProcessError as e:
        print("‚ùå Erro ao executar git pull:\n", e.output.decode())
        return Response(content="Erro ao atualizar o c√≥digo", status_code=500)
    except Exception as e:
        print("‚ùå Erro desconhecido:\n", str(e))
        return Response(content="Erro desconhecido", status_code=500)
